src/jailhouse/jailhouse_driver.c

```diff
@@ -1,35 +1,184 @@
+
+    return cfg;
+}
+
+static void virJailhouseDriverConfigDispose(void *obj)
+{
+
+    virJailhouseDriverConfigPtr cfg = obj;
+
+    VIR_FREE(cfg->sys_config_file_path);
+    VIR_FREE(cfg->cell_config_dir);
+}
+
+static int
+jailhouseLoadConf(virJailhouseDriverConfigPtr config)
+{
+    g_autoptr(virConf) conf = NULL;
+
+    if (!(conf = virConfReadFile(JAILHOUSE_CONFIG_FILE, 0)))
+        return -1;
+
+    if (virConfGetValueString(conf, "system_config",
+                              &config->sys_config_file_path) < 0)
+        return -1;
+
+    if (virConfGetValueString(conf, "non_root_cells_dir",
+                              &config->cell_config_dir) < 0)
+        return -1;
+
+    return 0;
+}
+
+static int
+jailhouseCreateAndLoadCells(virJailhouseDriverPtr driver)
+{
+    if (!driver->config ||
+        !driver->config->cell_config_dir ||
+        strlen(driver->config->cell_config_dir) == 0)
+        return -1;
+
+    // Create all cells in the hypervisor.
+    if (createJailhouseCells(driver->config->cell_config_dir) < 0)
+        return -1;
+
+    // Get all cells created above.
+    driver->cell_info_list = getJailhouseCellsInfo();
+
+    return 0;
+}
+
+static void
+jailhouseFreeDriver(virJailhouseDriverPtr driver)
+{
+    if (!driver)
+        return;
+
+    VIR_FREE(driver->config);
+    VIR_FREE(driver);
+}
 
 static virDrvOpenStatus
 jailhouseConnectOpen(virConnectPtr conn,
-                     virConnectAuthPtr auth,
-                     virConfPtr conf,
-                     unsigned int flags)
+                     virConnectAuthPtr auth G_GNUC_UNUSED,
+                     virConfPtr conf G_GNUC_UNUSED, unsigned int flags)
 {
-    UNUSED(conn);
-    UNUSED(auth);
-    UNUSED(conf);
-    UNUSED(flags);
-    return 0;
+    virJailhouseDriverPtr driver = NULL;
+    uid_t uid = geteuid();
+
+    virCheckFlags(VIR_CONNECT_RO, VIR_DRV_OPEN_ERROR);
+
+    if (!virConnectValidateURIPath(conn->uri->path, "jailhouse", uid == 0))
+        return VIR_DRV_OPEN_ERROR;
+
+    if (!virFileExists(JAILHOUSE_CONFIG_FILE)) {
+        virReportError(VIR_ERR_INTERNAL_ERROR,
+                       _("Jailhouse configuration file %s does not exist"),
+                       JAILHOUSE_CONFIG_FILE);
+        return VIR_DRV_OPEN_ERROR;
+    }
+
+    if (access(JAILHOUSE_CONFIG_FILE, R_OK) < 0) {
+        virReportError(VIR_ERR_INTERNAL_ERROR,
+                       _
+                       ("Jailhouse configuration file %s is not accessible"),
+                       JAILHOUSE_CONFIG_FILE);
+        return VIR_DRV_OPEN_ERROR;
+    }
+
+    if (VIR_ALLOC(driver) < 0)
+        return VIR_DRV_OPEN_ERROR;
+
+    if (!(driver->config = virJailhouseDriverConfigNew()))
+        goto cleanup;
+
+    if (jailhouseLoadConf(driver->config) < 0)
+        goto cleanup;
+
+    if (jailhouseEnable(driver->config->sys_config_file_path) < 0)
+        goto cleanup;
+
+    if (jailhouseCreateAndLoadCells(driver) < 0)
+        goto cleanup;
```

Comment:
Again, I feel that I touched on this previously, but if I didn't let's do that now.
This enables jailhouse and creates cells [1]. I might be misunderstanding but to me this is just a debug code, right? I mean, if I remember correctly, we discussed that jailhouse is a standalone hypervisor that doesn't need babysitting from libvirt, that it resembles virtualbox more than say qemu/uml/.. Or to ask differently - why do we need to enable jailhouse and individual cells here? To take this question even further - I don't fully understand the need for us to read *.cell files.

For instance, libxl (which is a driver for Xen) also uses config files which Xen is lated fed with. But it generates the config files from passed domain XML libxlBuildDomainConfig() and feed Xen with them. Similarly, this is how ESX works, on virDomainDefine() it parses the XML and generates a config file which is then passed to ESX. On virDomainGetXMLDesc() the flow is reversed - a configuration is obtained from ESX, which is then parsed and XML is constructed from it.

1: side question: what happens if those FDs passed to corresponding ioctl()-s are closed? Does cell continue to live? Does jailhouse stay enabled?

Reply:
I believe it was clear during the discussion in last week's meeting and I applied the patch you shared.

FDs are just used to read/write to Jailhouse device(/dev/jailhouse), so cell continues to live and jailhouse stays enabled even after FDs are closed.